Intro overhaul (#42)
oetiker/lshort: book/src/realworld.tex

\end{verse}
\end{example}

\section{Code Listings}

When writing about \LaTeX{} or other programming languages, you often need to
insert short code snippets like this \verb|\LaTeX{}|. While you could escape
all the characters, this would quickly become rather tiresome, especially for
longer pieces of code where correct spacing (collapsed by \LaTeX{} to single
spaces) is crucial for readability. We will present three solutions to this
problem.

\subsection{Verbatim}\label{sec:verbatim}

\LaTeX{} itself comes with the \ci{verb} command. It is unusual in that it does

> not use groups for getting its first argument. Instead you pass it between a

Reviewer: ```suggestion
not use groups for getting its first argument. Instead you pass the argument between a
```

> chosen delimiters that can be any character except for letters, \verb|*| or

Reviewer: ```suggestion
chosen delimiters that can be any character except a letter, \verb|*| or
```

space. For example to use \cargv{|} (which is typically used) as a delimiter
you type
\begin{lscommand}
  \ci{verb}\verb+|+\carg{text}\verb+|+
\end{lscommand}

> Text that is enclosed between these delimiters will be directly printed, as if

Reviewer: ```suggestion
Text that is enclosed between these delimiters will be printed without \LaTeX{} reacting to any special charaters or macros. The outpup will use the typewriter font.
```

\lstset{
  language=C,
  numbers=left,
  xleftmargin=1em, %!hide
  breaklines,
  basicstyle=\ttfamily,
  prebreak=\textrightarrow,
}

\lstinputlisting{hello.c}

The statement \lstinline|int x = 1;|
assigns 1 to variable \lstinline|x|.
\end{example}

\subsection{The \pai{minted} Package}

While the \pai{listings} package allows to typeset pretty decent looking code
snippets, you may find it still lacking in some respects. The most obvious
problem is the syntax highlighting: you may find that your language of choice
is not supported. You can write a custom style for the language but it relies
on a simple keyword search, so this is an option only for simple languages.

If this is a problem for you it may be worth to consider the \pai*{minted}
package. It relies on the Pygments~\cite{pygments} program to highlight code.
In contrast to the \pai{listings} package, this program uses pushdown automata
to color the code which may produce much better results. Moreover the
predefined styles and language support are much richer than those offered by
the \pai{listings} package. This is the package that is used to color examples
throughout this booklet.

> Due to the fact that the \pai{minted} package relies on third party program

Reviewer: ```suggestion
Due to the fact that the \pai{minted} package relies on an external program
```

there are two things that you need to do before using it
\begin{enumerate}
  \item Ensure that the Pygments program is installed. If you are using a
        \Unix{} system it is probably enough to execute\\
        \verb|pip install Pygments|\\
        You can find more information at their official website:
        \citeurl{pygments} and in the \pai*{minted} package documentation.

  \item
        Add \verb|--shell-escape| flag when compiling \LaTeX{} document, for
        example:\\
        \verb|xelatex --shell-escape document.tex|
\end{enumerate}
It is important to understand what the cryptic flag mentioned in the second

> point actually means. It allows the \LaTeX{} document to call any program via

Reviewer: ```suggestion
point actually means. It allows the \LaTeX{} document to call any external program via
```

> shell. Let me repeat this to make the point clear: \emph{Enabling this flag

Reviewer: ```suggestion
the shell. Let me repeat this to make the point clear: \emph{Enabling this flag
```

>   allows the \LaTeX{} document to call any program via shell}. This means that

Reviewer: ```suggestion
  allows the \LaTeX{} document to call any program via the shell}. This means that
```

the \cargv{breaksymbolleft} and \cargv{breaksymbolright} keys.

In order to use \LaTeX{} inside the comments you can use either the
\cargv{texcl} or its synonym \cargv{texcomments}. You can also add custom
\LaTeX{} escape delimiters using \cargv{escapeinside}, but unlike in
\pai{listings}, these must be single tokens.
\begin{example}
\begin{minted}[
  linenos,
  xleftmargin=1em, %!hide
  texcomments,
  escapeinside=||,
]{html}
<!DOCTYPE html>
<html>
  <head> |\label{head}|
    <title>On |\LaTeX{}|</title>
  </head>
  <body>
    <div class="review">
      <!-- \LaTeX{} review -->
      <p>It is awesome!</p>
    </div>
  </body>
</html>
\end{minted}
The \mintinline{html}{<head>} 
element in line~\ref{head}
contains page metadata.
\end{example}

The Pygments program comes with a lot of predefined styles for typesetting your
code. In order to choose between them use the \cargv{style} key.
\begin{example}
\inputminted[
  style=bw,
]{bash}{echo.sh}
\inputminted[
  style=sas,
]{bash}{echo.sh}
\inputminted[
  style=xcode,
]{bash}{echo.sh}
\end{example}

To avoid setting the options with every listing with every code snippet you

> include you can use \ci{setminted} to define the default values.

Reviewer: ```suggestion
include, you can use \ci{setminted} to define the default values.
```
